File: C++/2-Medium/00173._二叉搜索树迭代器/00173._二叉搜索树迭代器.cpp
#include <iostream>
#include <stack>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class BSTIterator {
private:
    stack<TreeNode*> nodeStack; // 栈用于存储中序遍历的状态

    // 将给定节点及其所有左子节点压入栈中
    void pushLeft(TreeNode* node) {
        while (node != nullptr) {
            nodeStack.push(node);
            node = node->left;
        }
    }

public:
    BSTIterator(TreeNode* root) {
        pushLeft(root); // 初始化时，将根节点及其所有左子节点压入栈中
    }
    
    // 返回二叉搜索树中的下一个最小的数
    int next() {
        TreeNode* nextNode = nodeStack.top();
        nodeStack.pop();
        pushLeft(nextNode->right); // 将下一个节点的右子节点及其所有左子节点压入栈中
        return nextNode->val;
    }
    
    // 返回二叉搜索树中是否还有下一个数
    bool hasNext() {
        return !nodeStack.empty();
    }
};

// 用于测试的主函数
int main() {
    // 示例树的构建
    TreeNode* root = new TreeNode(7);
    root->left = new TreeNode(3);
    root->right = new TreeNode(15);
    root->right->left = new TreeNode(9);
    root->right->right = new TreeNode(20);

    BSTIterator iterator(root);
    cout << iterator.next() << " ";    // 输出 3
    cout << iterator.next() << " ";    // 输出 7
    cout << (iterator.hasNext() ? "true" : "false") << " "; // 输出 true
    cout << iterator.next() << " ";    // 输出 9
    cout << (iterator.hasNext() ? "true" : "false") << " "; // 输出 true
    cout << iterator.next() << " ";    // 输出 15
    cout << (iterator.hasNext() ? "true" : "false") << " "; // 输出 true
    cout << iterator.next() << " ";    // 输出 20
    cout << (iterator.hasNext() ? "true" : "false") << endl; // 输出 false

    return 0;
}

File: Python/2-Medium/00173._二叉搜索树迭代器/00173._二叉搜索树迭代器.py
from typing import Optional, Generator

class TreeNode:
    def __init__(self, x: int, left: 'TreeNode' = None, right: 'TreeNode' = None):
        self.val = x
        self.left = left
        self.right = right

class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        """
        初始化二叉搜索树迭代器，使用栈来存储从根节点到最左侧节点的路径。
        
        :param root: 二叉搜索树的根节点
        """
        self.stack = []
        self._leftmost_inorder(root)
    
    def _leftmost_inorder(self, root: Optional[TreeNode]):
        """
        将给定节点到最左侧节点的路径推入栈中。
        
        :param root: 当前节点
        """
        while root:
            self.stack.append(root)
            root = root.left
    
    def next(self) -> int:
        """
        返回二叉搜索树中的下一个最小的数。
        
        :return: 下一个最小的数
        """
        # 弹出栈顶元素，即当前的最小节点
        topmost_node = self.stack.pop()
        # 如果这个节点有右子节点，则将右子节点及其所有左子节点推入栈中
        if topmost_node.right:
            self._leftmost_inorder(topmost_node.right)
        return topmost_node.val
    
    def hasNext(self) -> bool:
        """
        返回二叉搜索树中是否还有下一个更小的数。
        
        :return: 如果还有下一个最小的数，则返回True；否则返回False
        """
        return len(self.stack) > 0

# 构建示例二叉搜索树
root = TreeNode(7, TreeNode(3), TreeNode(15, TreeNode(9), TreeNode(20)))

# 初始化迭代器并演示其功能
iterator = BSTIterator(root)
operations = ['next', 'hasNext', 'next', 'hasNext', 'next', 'hasNext', 'next', 'hasNext', 'next', 'hasNext']
results = []

for op in operations:
    if op == 'next':
        results.append(iterator.next())
    elif op == 'hasNext':
        results.append(iterator.hasNext())

print(results)  # [3, True, 7, True, 9, True, 15, True, 20, False]